File: app/src/main/java/org/neidhardt/dynamicsoundboard/navigationdrawer/NavigationDrawerFragment.java
package org.neidhardt.dynamicsoundboard.navigationdrawer;

import android.animation.Animator;
import android.content.res.Resources;
import android.os.Bundle;
import android.support.annotation.Nullable;
import android.support.v4.view.PagerAdapter;
import android.support.v4.view.ViewPager;
import android.support.v7.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.view.animation.DecelerateInterpolator;
import android.widget.TextView;
import de.greenrobot.event.EventBus;
import org.neidhardt.dynamicsoundboard.DynamicSoundboardApplication;
import org.neidhardt.dynamicsoundboard.R;
import org.neidhardt.dynamicsoundboard.dialog.AddNewSoundSheetDialog;
import org.neidhardt.dynamicsoundboard.dialog.addnewsound.AddNewSoundDialog;
import org.neidhardt.dynamicsoundboard.dialog.soundlayouts.AddNewSoundLayoutDialog;
import org.neidhardt.dynamicsoundboard.misc.AnimationUtils;
import org.neidhardt.dynamicsoundboard.navigationdrawer.events.ActionModeEvent;
import org.neidhardt.dynamicsoundboard.navigationdrawer.playlist.Playlist;
import org.neidhardt.dynamicsoundboard.navigationdrawer.playlist.PlaylistAdapter;
import org.neidhardt.dynamicsoundboard.navigationdrawer.soundsheets.SoundSheets;
import org.neidhardt.dynamicsoundboard.navigationdrawer.soundsheets.SoundSheetsAdapter;
import org.neidhardt.dynamicsoundboard.soundactivity.BaseFragment;
import org.neidhardt.dynamicsoundboard.soundlayouts.SoundLayoutsList;
import org.neidhardt.dynamicsoundboard.soundlayouts.SoundLayoutsListAdapter;
import org.neidhardt.dynamicsoundboard.soundlayouts.SoundLayoutsManager;
import org.neidhardt.dynamicsoundboard.soundsheetmanagement.SoundSheetsManagerFragment;

public class NavigationDrawerFragment
		extends
			BaseFragment
		implements
			View.OnClickListener,
			ViewPager.OnPageChangeListener
{
	public static final String TAG = NavigationDrawerFragment.class.getName();

	private static final int INDEX_SOUND_SHEETS = 0;
	private static final int INDEX_PLAYLIST = 1;

	private EventBus bus;

	private SlidingTabLayout tabBar;
	private ViewPager tabContent;
	private TabContentAdapter tabContentAdapter;

	private ViewGroup listContainer;
	private ViewPagerContentObserver listObserver;
	private SoundLayoutsList soundLayoutList;
	private SoundLayoutsListAdapter soundLayoutListAdapter;
	private Playlist playlist;
	private PlaylistAdapter playlistAdapter;
	private SoundSheets soundSheets;
	private TextView currentLayoutName;

	private View contextualActionContainer;
	private View deleteSelected;

	private int minHeightOfListContent = 0;

	@Override
	public void onCreate(Bundle savedInstanceState)
	{
		super.onCreate(savedInstanceState);
		this.setRetainInstance(true);

		this.listObserver = new ViewPagerContentObserver();
		this.tabContentAdapter = new TabContentAdapter();
		this.soundLayoutListAdapter = new SoundLayoutsListAdapter();
		this.playlistAdapter = new PlaylistAdapter();

		this.bus = EventBus.getDefault();
	}

	@Nullable
	@Override
	public View onCreateView(LayoutInflater inflater, ViewGroup container, Bundle savedInstanceState)
	{
		View fragmentView = inflater.inflate(R.layout.fragment_navigation_drawer, container, false);

		this.contextualActionContainer = fragmentView.findViewById(R.id.layout_contextual_controls);
		this.listContainer = (ViewGroup) fragmentView.findViewById(R.id.layout_navigation_drawer_list_content);

		this.deleteSelected = fragmentView.findViewById(R.id.b_delete_selected);
		this.deleteSelected.setOnClickListener(this);

		fragmentView.findViewById(R.id.b_delete).setOnClickListener(this);
		fragmentView.findViewById(R.id.b_ok).setOnClickListener(this);
		fragmentView.findViewById(R.id.layout_change_sound_layout).setOnClickListener(this);

		this.tabContent = (ViewPager) fragmentView.findViewById(R.id.vp_tab_content);
		this.tabContent.setAdapter(this.tabContentAdapter);
		this.tabContent.setOnPageChangeListener(this);

		this.tabBar = (SlidingTabLayout) fragmentView.findViewById(R.id.layout_tab);
		this.tabBar.setOnPageChangeListener(this);
		this.tabBar.setViewPager(tabContent);
		this.tabBar.setCustomTabColorizer(new NavigationDrawerTabColorizer());

		this.soundLayoutList = (SoundLayoutsList) fragmentView.findViewById(R.id.layout_select_sound_layout);
		this.soundLayoutList.setAdapter(this.soundLayoutListAdapter);

		this.playlist = (Playlist) fragmentView.findViewById(R.id.playlist);
		this.playlist.setAdapter(this.playlistAdapter);
		this.initPlayListAndAdapter();

		this.soundSheets = (SoundSheets) fragmentView.findViewById(R.id.sound_sheets);
		this.initSoundSheetsAndAdapter();

		return fragmentView;
	}

	@Override
	public void onActivityCreated(Bundle savedInstanceState)
	{
		super.onActivityCreated(savedInstanceState);

		this.currentLayoutName = (TextView) this.getActivity().findViewById(R.id.tv_current_sound_layout_name);
		this.currentLayoutName.setText(SoundLayoutsManager.getInstance().getActiveSoundLayout().getLabel());
	}

	@Override
	public void onResume()
	{
		super.onResume();

		this.initSoundLayoutsAndAdapter();
		this.initSoundSheetsAndAdapter();
		this.initPlayListAndAdapter();

		this.calculateMinHeightOfListContent();
		this.adjustViewPagerToContent();

		this.playlistAdapter.registerAdapterDataObserver(this.listObserver);
		this.soundSheets.getAdapter().registerAdapterDataObserver(this.listObserver);
	}

	/**
	 * Calculates the minimum require height of the viewpager's content (this is the height used if the content is smaller than the
	 * screens height). Recalculation is require every time the screen's metric changes (ie. switch from/to full immersive mode).
	 */
	public void calculateMinHeightOfListContent()
	{
		this.minHeightOfListContent = this.contextualActionContainer.getTop() - listContainer.getTop();  // this is the minimal height required to fill the screen properly
	}

	private void initSoundLayoutsAndAdapter()
	{
		this.soundLayoutList.setParentFragment(this);
		this.soundLayoutListAdapter.setNavigationDrawerFragment(this);
	}

	private void initSoundSheetsAndAdapter()
	{
		this.soundSheets.setParentFragment(this);
		this.soundSheets.getAdapter().setNavigationDrawerFragment(this);
		this.soundSheets.getAdapter().notifyDataSetChanged();
	}

	private void initPlayListAndAdapter()
	{
		this.playlist.setParentFragment(this);
		this.playlistAdapter.setServiceManagerFragment(this.getServiceManagerFragment());
		this.playlistAdapter.startProgressUpdateTimer();
		if (!EventBus.getDefault().isRegistered(this.playlistAdapter))
			EventBus.getDefault().register(this.playlistAdapter);
	}

	@Override
	public void onStart()
	{
		super.onStart();
		if (!this.bus.isRegistered(this.playlistAdapter))
			this.bus.register(this.playlistAdapter);
		if (!this.bus.isRegistered(this))
			this.bus.register(this);
	}

	@Override
	public void onStop()
	{
		super.onStop();
		this.bus.unregister(this.playlistAdapter);
		this.bus.unregister(this);
	}

	@Override
	public void onPause()
	{
		super.onPause();

		EventBus.getDefault().unregister(this.playlistAdapter);
		this.playlistAdapter.unregisterAdapterDataObserver(this.listObserver);
		this.soundSheets.getAdapter().unregisterAdapterDataObserver(this.listObserver);

		this.playlistAdapter.stopProgressUpdateTimer();
		this.playlist.setParentFragment(null);

		this.soundSheets.setParentFragment(null);
	}

	@Override
	public void onClick(View v)
	{
		int id = v.getId();
		if (id == R.id.b_delete)
		{
			if (this.soundLayoutList.isActive())
				this.soundLayoutList.prepareItemDeletion();
			else if (this.tabContent.getCurrentItem() == INDEX_PLAYLIST)
				this.playlist.prepareItemDeletion();
			else
				this.soundSheets.prepareItemDeletion();
		}
		else if (id == R.id.b_delete_selected)
		{
			if (this.soundLayoutList.isActive())
				this.soundLayoutList.deleteSelected();
			else if (this.tabContent.getCurrentItem() == INDEX_PLAYLIST)
				this.playlist.deleteSelected();
			else
				this.soundSheets.deleteSelected();
		}
		else if (id  == R.id.b_ok)
		{
			if (this.soundLayoutList.isActive())
				AddNewSoundLayoutDialog.showInstance(this.getFragmentManager(), SoundLayoutsManager.getInstance().getSuggestedSoundLayoutName());
			else if (this.tabContent.getCurrentItem() == INDEX_PLAYLIST)
				AddNewSoundDialog.showInstance(this.getFragmentManager(), Playlist.TAG);
			else
			{
				SoundSheetsManagerFragment fragment = this.getSoundSheetManagerFragment();
				AddNewSoundSheetDialog.showInstance(this.getFragmentManager(), fragment.getSuggestedSoundSheetName());
			}
		}
		else if (id == R.id.layout_change_sound_layout)
		{
			this.animateSoundLayoutsListAppear();
			this.soundLayoutList.toggleVisibility();
			if (this.getBaseActivity().isActionModeActive() && this.soundLayoutList.isActive())
				this.soundLayoutList.prepareItemDeletion();
		}
	}

	private void animateSoundLayoutsListAppear()
	{
		View indicator = this.getActivity().findViewById(R.id.iv_change_sound_layout_indicator);
		indicator.animate()
				.rotationXBy(180)
				.setDuration(this.getResources().getInteger(android.R.integer.config_shortAnimTime))
				.start();

		final View viewToAnimate = this.getActivity().findViewById(R.id.v_reveal_shadow);
		Animator animator = AnimationUtils.createSlowCircularReveal(viewToAnimate,
				this.listContainer.getWidth(), 0,
				0, 2 * this.listContainer.getHeight());

		if (animator != null)
			animator.start();
	}

	public void triggerSoundLayoutUpdate()
	{
		if (this.currentLayoutName != null)
			this.currentLayoutName.setText(SoundLayoutsManager.getInstance().getActiveSoundLayout().getLabel());
		this.soundLayoutListAdapter.notifyDataSetChanged();
	}

	/**
	 * This is called by greenRobot EventBus in case a request to change the current contextual action mode has benn submitted.
	 * playlist entries.
	 * @param event delivered OpenSoundSheetEvent
	 */
	@SuppressWarnings("unused")
	public void onEvent(ActionModeEvent event)
	{
		ActionModeEvent.REQUEST requestedAction = event.getRequestedAction();
		switch (requestedAction)
		{
			case START:
				this.onActionModeStart();
				return;
			case STOPPED:
				this.onActionModeFinished();
		}
	}

	private void onActionModeStart()
	{
		this.deleteSelected.setVisibility(View.VISIBLE);
		int distance = this.contextualActionContainer.getWidth();

		this.deleteSelected.setTranslationX(-distance);
		this.deleteSelected.animate().
				translationX(0).
				setDuration(this.getResources().getInteger(android.R.integer.config_mediumAnimTime)).
				setInterpolator(new DecelerateInterpolator()).
				start();
	}

	private void onActionModeFinished()
	{
		this.deleteSelected.setVisibility(View.GONE);
	}

	@Override
	public void onPageScrolled(int position, float positionOffset, int positionOffsetPixels) {}

	@Override
	public void onPageSelected(int position)
	{
		if (!this.getBaseActivity().isActionModeActive())
			return;
		if (position == INDEX_SOUND_SHEETS)
			this.soundSheets.prepareItemDeletion();
		else if (position == INDEX_PLAYLIST)
			this.playlist.prepareItemDeletion();
	}

	@Override
	public void onPageScrollStateChanged(int state) {}

	public void setLayoutName(String layoutName)
	{
		if (this.currentLayoutName != null)
			this.currentLayoutName.setText(layoutName);
	}

	/**
	 * This function resize the view pagers height to its content. It is necessary, because the viewpager can not
	 * have layout parameter wrap_content.
	 */
	public void adjustViewPagerToContent()
	{
		Resources resources = DynamicSoundboardApplication.getSoundboardContext().getResources();
		int childHeight = resources.getDimensionPixelSize(R.dimen.height_list_item);
		int dividerHeight = resources.getDimensionPixelSize(R.dimen.stroke);
		int padding = resources.getDimensionPixelSize(R.dimen.margin_small);

		int soundSheetCount = this.getSoundSheetManagerFragment().getSoundSheets().size();
		int playListCount = this.playlistAdapter.getItemCount();

		int heightSoundSheetChildren = soundSheetCount * childHeight;
		int heightDividerSoundSheet = soundSheetCount > 1 ? (soundSheetCount - 1) * dividerHeight : 0;
		int heightSoundSheet = heightSoundSheetChildren + heightDividerSoundSheet + padding + this.tabBar.getHeight();

		int heightPlayListChildren = playListCount * childHeight;
		int heightDividerPlayList = playListCount > 1 ? (playListCount - 1) * dividerHeight : 0;
		int heightPlayList = heightPlayListChildren + heightDividerPlayList + padding + this.tabBar.getHeight();

		int largestList = Math.max(heightSoundSheet, heightPlayList);
		if (this.minHeightOfListContent == 0) // 0 means the current height was not measured, remeasure
			this.minHeightOfListContent = this.contextualActionContainer.getTop() - listContainer.getTop();

		this.listContainer.getLayoutParams().height = Math.max(largestList, minHeightOfListContent);
	}

	public SoundSheetsAdapter getSoundSheetsAdapter()
	{
		return this.soundSheets.getAdapter();
	}

	public Playlist getPlaylist()
	{
		return this.playlist;
	}

	private class TabContentAdapter extends PagerAdapter
	{
		@Override
		public CharSequence getPageTitle(int position)
		{
			if (position == INDEX_SOUND_SHEETS)
				return getResources().getString(R.string.tab_sound_sheets);
			else
				return getResources().getString(R.string.tab_play_list);
		}

		@Override
		public int getCount()
		{
			return 2;
		}

		@Override
		public boolean isViewFromObject(View view, Object object)
		{
			return view.equals(object);
		}

		@Override
		public Object instantiateItem(ViewGroup container, int position)
		{
			switch (position)
			{
				case INDEX_SOUND_SHEETS:
					return soundSheets;
				case INDEX_PLAYLIST:
					return playlist;
				default:
					throw new NullPointerException("instantiateItem: no view for position " + position + " is available");
			}
		}
	}

	private class ViewPagerContentObserver extends RecyclerView.AdapterDataObserver
	{
		@Override
		public void onChanged()
		{
			super.onChanged();
			adjustViewPagerToContent();
		}
	}

	private class NavigationDrawerTabColorizer implements SlidingTabLayout.TabColorizer
	{
		@Override
		public int getIndicatorColor(int position)
		{
			return getResources().getColor(R.color.accent_200);
		}

		@Override
		public int getDividerColor(int position)
		{
			return 0;
		}
	}
}

File: app/src/main/kotlin/org/neidhardt/dynamicsoundboard/navigationdrawer/NavigationDrawerFragment.kt
package org.neidhardt.dynamicsoundboard.navigationdrawer

import android.app.FragmentManager
import android.os.Bundle
import android.support.design.widget.AppBarLayout
import android.support.design.widget.TabLayout
import android.support.v7.widget.DefaultItemAnimator
import android.support.v7.widget.LinearLayoutManager
import android.support.v7.widget.RecyclerView
import android.support.v7.widget.Toolbar
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.view.animation.DecelerateInterpolator
import org.greenrobot.eventbus.EventBus
import org.greenrobot.eventbus.Subscribe
import org.greenrobot.eventbus.ThreadMode
import org.neidhardt.dynamicsoundboard.R
import org.neidhardt.dynamicsoundboard.SoundboardApplication
import org.neidhardt.dynamicsoundboard.mediaplayer.PlaylistTAG
import org.neidhardt.dynamicsoundboard.misc.AnimationUtils
import org.neidhardt.dynamicsoundboard.navigationdrawer.header.events.OnOpenSoundLayoutsEventListener
import org.neidhardt.dynamicsoundboard.navigationdrawer.header.events.OpenSoundLayoutsRequestedEvent
import org.neidhardt.dynamicsoundboard.navigationdrawer.playlist.PlaylistAdapter
import org.neidhardt.dynamicsoundboard.navigationdrawer.playlist.PlaylistPresenter
import org.neidhardt.dynamicsoundboard.navigationdrawer.soundlayouts.SoundLayoutsAdapter
import org.neidhardt.dynamicsoundboard.navigationdrawer.soundlayouts.SoundLayoutsPresenter
import org.neidhardt.dynamicsoundboard.navigationdrawer.soundsheets.SoundSheetsAdapter
import org.neidhardt.dynamicsoundboard.navigationdrawer.soundsheets.SoundSheetsPresenter
import org.neidhardt.dynamicsoundboard.soundactivity.BaseFragment
import org.neidhardt.dynamicsoundboard.soundlayoutmanagement.events.OnSoundLayoutSelectedEventListener
import org.neidhardt.dynamicsoundboard.soundlayoutmanagement.events.SoundLayoutSelectedEvent
import org.neidhardt.dynamicsoundboard.soundlayoutmanagement.model.SoundLayoutsAccess
import org.neidhardt.dynamicsoundboard.soundlayoutmanagement.model.SoundLayoutsStorage
import org.neidhardt.dynamicsoundboard.soundlayoutmanagement.model.SoundLayoutsUtil
import org.neidhardt.dynamicsoundboard.soundlayoutmanagement.views.AddNewSoundLayoutDialog
import org.neidhardt.dynamicsoundboard.soundmanagement.dialog.AddNewSoundDialog
import org.neidhardt.dynamicsoundboard.soundmanagement.model.SoundsDataAccess
import org.neidhardt.dynamicsoundboard.soundmanagement.model.SoundsDataStorage
import org.neidhardt.dynamicsoundboard.soundsheetmanagement.model.SoundSheetsDataAccess
import org.neidhardt.dynamicsoundboard.soundsheetmanagement.model.SoundSheetsDataStorage
import org.neidhardt.dynamicsoundboard.soundsheetmanagement.model.SoundSheetsDataUtil
import org.neidhardt.dynamicsoundboard.soundsheetmanagement.views.AddNewSoundSheetDialog
import org.neidhardt.dynamicsoundboard.views.recyclerviewhelpers.DividerItemDecoration

class NavigationDrawerFragment : BaseFragment()
{
	private val eventBus = EventBus.getDefault()
	private var presenter: NavigationDrawerFragmentPresenter? = null

	override fun onCreate(savedInstanceState: Bundle?)
	{
		super.onCreate(savedInstanceState)
		this.retainInstance = true
	}

	override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View?
	{
		val view = inflater.inflate(R.layout.fragment_navigation_drawer, container, false)

		val tabLayout = (view.findViewById(R.id.tl_tab_bar) as TabLayout).apply {
			this.tabGravity = TabLayout.GRAVITY_FILL
		}

		val layoutList = (view.findViewById(R.id.rv_navigation_drawer_list) as RecyclerView).apply {
			this.itemAnimator = DefaultItemAnimator()
			this.layoutManager = LinearLayoutManager(this.context)
			this.addItemDecoration(DividerItemDecoration(this.context))
		}

		this.presenter = NavigationDrawerFragmentPresenter(
				eventBus = this.eventBus,

				toolbar = view.findViewById(R.id.toolbar_navigation_drawer) as Toolbar,
				appBarLayout = view.findViewById(R.id.abl_navigation_drawer) as AppBarLayout,
				tabLayout = tabLayout,
				buttonOk = view.findViewById(R.id.b_ok),
				buttonDelete = view.findViewById(R.id.b_delete),
				buttonDeleteSelected = view.findViewById(R.id.b_delete_selected),
				revealShadow = view.findViewById(R.id.v_reveal_shadow),

				fragmentManager = this.fragmentManager,
				recyclerView = layoutList,

				soundsDataAccess = SoundboardApplication.getSoundsDataAccess(),
				soundsDataStorage = SoundboardApplication.getSoundsDataStorage(),

				soundSheetsDataUtil = SoundboardApplication.getSoundSheetsDataUtil(),
				soundSheetsDataStorage = SoundboardApplication.getSoundSheetsDataStorage(),
				soundSheetsDataAccess = SoundboardApplication.getSoundSheetsDataAccess(),

				soundLayoutsAccess = SoundboardApplication.getSoundLayoutsAccess(),
				soundLayoutsStorage = SoundboardApplication.getSoundLayoutsStorage(),
				soundLayoutsUtil = SoundboardApplication.getSoundLayoutsUtil()

		).apply {
			onAttachedToWindow()
		}

		return view
	}

	override fun onStart()
	{
		super.onStart()
		this.presenter?.onAttachedToWindow()
	}

	override fun onStop() {
		super.onStop()
		this.presenter?.onDetachedFromWindow()
	}
}

enum class List
{
	SoundSheet,
	Playlist,
	SoundLayouts
}

private val INDEX_SOUND_SHEETS = 0
private val INDEX_PLAYLIST = 1

class NavigationDrawerFragmentPresenter
(
		private val eventBus: EventBus,
		private val fragmentManager: FragmentManager,

		private val toolbar: Toolbar,
		private val appBarLayout: AppBarLayout,
		private val tabLayout: TabLayout,
		private val revealShadow: View,
		private val buttonOk: View,
		private val buttonDelete: View,
		private val buttonDeleteSelected: View,

		private val recyclerView: RecyclerView,

		private val soundLayoutsAccess: SoundLayoutsAccess,
		private val soundLayoutsStorage: SoundLayoutsStorage,
		private val soundLayoutsUtil: SoundLayoutsUtil,

		private val soundsDataAccess: SoundsDataAccess,
		private val soundsDataStorage: SoundsDataStorage,

		private val soundSheetsDataAccess: SoundSheetsDataAccess,
		private val soundSheetsDataStorage: SoundSheetsDataStorage,
		private val soundSheetsDataUtil: SoundSheetsDataUtil

) :
		View.OnClickListener,
		OnOpenSoundLayoutsEventListener,
		TabLayout.OnTabSelectedListener,
		OnSoundLayoutSelectedEventListener
{
	private var tabSoundSheets: TabLayout.Tab = tabLayout.createSoundSheetTab()
	private var tabPlayList: TabLayout.Tab = tabLayout.createPlaylistTab()
	private var tabSoundLayouts: TabLayout.Tab = tabLayout.createSoundLayoutsTab()

	private var currentList: List = List.SoundSheet
	private var currentPresenter: NavigationDrawerListPresenter? = null

	private val presenterSoundSheets = createSoundSheetPresenter(eventBus, soundsDataAccess, soundsDataStorage, soundSheetsDataAccess, soundSheetsDataStorage)
	private val presenterPlaylist = createPlaylistPresenter(eventBus, soundsDataAccess, soundsDataStorage)
	private val presenterSoundLayouts = createSoundLayoutsPresenter(eventBus, soundLayoutsAccess, soundLayoutsStorage)

	init
	{
		this.tabLayout.setOnTabSelectedListener(this)
		this.buttonOk.setOnClickListener(this)
		this.buttonDelete.setOnClickListener(this)
		this.buttonDeleteSelected.setOnClickListener(this)
	}

	fun onAttachedToWindow()
	{
		this.showDefaultTabBarAndContent()

		if (!this.eventBus.isRegistered(this))
			this.eventBus.register(this)
	}

	fun onDetachedFromWindow(): Unit = this.eventBus.unregister(this)

	private fun showDefaultTabBarAndContent()
	{
		this.tabLayout.removeAllTabs()
		this.tabSoundSheets = this.tabLayout.createSoundSheetTab()
		this.tabPlayList = this.tabLayout.createPlaylistTab()

		this.tabLayout.addTab(this.tabSoundSheets, INDEX_SOUND_SHEETS)
		this.tabLayout.addTab(this.tabPlayList, INDEX_PLAYLIST)

		this.tabSoundSheets.select()
	}

	private fun showContextTabBarAndContent()
	{
		this.tabLayout.removeAllTabs()
		this.tabSoundLayouts = this.tabLayout.createSoundLayoutsTab()

		this.tabLayout.addTab(this.tabSoundLayouts)
		this.tabSoundLayouts.select()
	}

	private fun showToolbarForDeletion()
	{
		this.toolbar.visibility = View.VISIBLE
		this.appBarLayout.setExpanded(false, true)
		this.recyclerView.isNestedScrollingEnabled = false

		val distance = this.recyclerView.width
		this.buttonDeleteSelected.apply {
			this.visibility = View.VISIBLE
			this.translationX = (-distance).toFloat()
			animate()
					.translationX(0f)
					.setDuration(this.resources.getInteger(android.R.integer.config_mediumAnimTime).toLong())
					.setInterpolator(DecelerateInterpolator())
					.start()
		}
	}

	private fun hideToolbarForDeletion()
	{
		this.toolbar.visibility = View.GONE
		this.appBarLayout.setExpanded(true, true)
		this.recyclerView.isNestedScrollingEnabled = false

		this.buttonDeleteSelected.visibility = View.GONE
	}

	override fun onClick(view: View)
	{
		val id = view.id
		when (id)
		{
			this.buttonDelete.id ->
			{
				this.showToolbarForDeletion()
				this.currentPresenter?.startDeletionMode()
			}
			this.buttonDeleteSelected.id ->
			{
				this.currentPresenter?.deleteSelectedItems()
				this.currentPresenter?.stopDeletionMode()
				this.hideToolbarForDeletion()
			}
			this.buttonOk.id ->
				if (this.currentList == List.SoundLayouts)
				{
					AddNewSoundLayoutDialog.showInstance(this.fragmentManager, this.soundLayoutsUtil.getSuggestedName())
				}
				else if (this.currentList == List.Playlist)
				{
					AddNewSoundDialog(this.fragmentManager, PlaylistTAG)
				}
				else if (this.currentList == List.SoundSheet)
				{
					AddNewSoundSheetDialog.showInstance(this.fragmentManager, this.soundSheetsDataUtil.getSuggestedName())
				}
		}
	}

	override fun onTabSelected(tab: TabLayout.Tab?)
	{
		this.currentPresenter?.onDetachedFromWindow()

		when (tab)
		{
			this.tabSoundSheets ->
			{
				this.currentList = List.SoundSheet
				this.currentPresenter = this.presenterSoundSheets
				this.recyclerView.adapter = this.presenterSoundSheets.adapter
			}
			this.tabPlayList ->
			{
				this.currentList = List.Playlist
				this.currentPresenter = this.presenterPlaylist
				this.recyclerView.adapter = this.presenterPlaylist.adapter
			}
			this.tabSoundLayouts ->
			{
				this.currentList = List.SoundLayouts
				this.currentPresenter = this.presenterSoundLayouts
				this.recyclerView.adapter = this.presenterSoundLayouts.adapter
			}
		}

		this.currentPresenter?.onAttachedToWindow()
	}

	@Subscribe(threadMode = ThreadMode.MAIN)
	override fun onEvent(event: OpenSoundLayoutsRequestedEvent)
	{
		if (event.openSoundLayouts) {
			this.showContextTabBarAndContent()
			this.animateSoundLayoutsListAppear()
		}
		else {
			this.showDefaultTabBarAndContent()
			this.animateSoundLayoutsListAppear()
		}
	}

	@Subscribe(threadMode = ThreadMode.MAIN)
	override fun onEvent(event: SoundLayoutSelectedEvent)
	{
		this.showDefaultTabBarAndContent()
	}

	private fun animateSoundLayoutsListAppear()
	{
		if (this.revealShadow.isAttachedToWindow)
		{
			val animator = AnimationUtils.createSlowCircularReveal(this.revealShadow, this.recyclerView.width, 0, 0f, (2 * this.recyclerView.height).toFloat())
			animator?.start()
		}
	}

	override fun onTabReselected(tab: TabLayout.Tab?) {}
	override fun onTabUnselected(tab: TabLayout.Tab?) {}
}

private fun createSoundSheetPresenter(
		eventBus: EventBus,
		soundsDataAccess: SoundsDataAccess, soundsDataStorage: SoundsDataStorage,
		soundSheetsDataAccess: SoundSheetsDataAccess, soundSheetsDataStorage: SoundSheetsDataStorage): SoundSheetsPresenter
{
	return SoundSheetsPresenter(
			eventBus = eventBus,
			soundsDataAccess = soundsDataAccess,
			soundsDataStorage = soundsDataStorage,
			soundSheetsDataAccess = soundSheetsDataAccess,
			soundSheetsDataStorage = soundSheetsDataStorage
	).apply {
		this.adapter = SoundSheetsAdapter(this)
	}
}

private fun createPlaylistPresenter(
		eventBus: EventBus, soundsDataAccess: SoundsDataAccess, soundsDataStorage: SoundsDataStorage): PlaylistPresenter
{
	return PlaylistPresenter(
			eventBus = eventBus,
			soundsDataAccess = soundsDataAccess,
			soundsDataStorage = soundsDataStorage
	).apply {
		this.adapter = PlaylistAdapter(this)
	}
}

private fun createSoundLayoutsPresenter(
		eventBus: EventBus, soundLayoutsAccess: SoundLayoutsAccess, soundLayoutsStorage: SoundLayoutsStorage): SoundLayoutsPresenter
{
	return SoundLayoutsPresenter(
			eventBus = eventBus,
			soundLayoutsAccess = soundLayoutsAccess,
			soundLayoutsStorage = soundLayoutsStorage
	).apply {
		this.adapter = SoundLayoutsAdapter(eventBus, this)
	}
}

private fun TabLayout.createSoundSheetTab(): TabLayout.Tab = this.newTab().setText(R.string.tab_sound_sheets)

private fun TabLayout.createPlaylistTab(): TabLayout.Tab = this.newTab().setText(R.string.tab_play_list)

private fun TabLayout.createSoundLayoutsTab(): TabLayout.Tab = this.newTab().setText(R.string.navigation_drawer_select_sound_layout)
